File: reference/Player.java
import java.util.ArrayList;
import java.util.Optional;
import java.io.File;

import javafx.scene.Scene;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.paint.Color;
import javafx.scene.effect.Light;
import javafx.scene.effect.Lighting;
import javafx.scene.input.KeyCode;
import javafx.scene.image.Image;
import javafx.scene.media.Media;
import javafx.scene.media.MediaPlayer;
import javafx.scene.media.AudioClip;

public class Player
{
	private double x; //of top-left corner
	private double xi; //initial x position (reset when velocity is reset)
	private double y; //of top-left corner
	private double yi; //initial y position (reset when velocity is reset)
	private double vx;
	private double vy;
	private double ay;
	
	private double vy_boost; //extra vy to gain from pressing the down key
	private double jump_speed;
	private double horiz_speed;
	
	private double time; //time of most recent updatePosition call
	private double ti_x; //time of most recent x velocity reset
	private double ti_y; //time of most recent y velocity reset
	private double t_hit; //time the player most recently got hit, used for making them red when hit
	
	private int x_collision; // -1 means wall to left, 0 means no collision, 1 means wall to right
	private int y_collision; // -1 means wall above, 0 means no collision, 1 means wall below
	private boolean border_y_collision; //set to true if the y_collision is with a border wall
	
	private Image hat;
	private Image rightChungus;
	private Image leftChungus;
	private double width; //width of hitbox and of chungus image
	private double chungus_height; //height of chungus image
	private double height; //height of hitbox
	
	private int health;
	private double t_threw_ball; //time when the player last threw a ball; players are invincible for a certain timeout after throwing a ball
	private double invincible_timeout; //this is that timeout, in sec
	private double red_color_timeout; //how long to make the player red after getting hit, in sec
	
	private KeyCode jumpKey;
	private KeyCode leftKey;
	private KeyCode rightKey;
	private KeyCode downKey;
	private KeyCode rotateLeftKey; //for launch angle changing
	private KeyCode ballKey;
	private KeyCode rotateRightKey;
	private boolean downKeyProcessed; //keeps track if we processed the first down key press, used so that we only process the first one
	
	private String direction; //"left" or "right"; direction the player is facing
	
	private Optional<Ball> myBall;
	private double[] shootAngleArray; //stores possible shoot angles, from 0 to pi/2.  These are positive when shooting up as it appears on the canvas.
	private int shootAngleIndex; //index in shoot angle array to reference
	private double shootAngle; //in radians, actual angle of launch, in math coords, not canvas coords
	private double shootVelocity; //in px/s
	
	private ArrayList<Wall> walls; //will be equal to the main array of these
	private ArrayList<Ball> balls;
	
	public Player(double x,double y,double width, Image hat, double ay)
	{
		this.x = x;
		this.xi = x;
		this.y = y;
		this.yi = y;
		this.vx = 0;
		this.vy = 0;
		this.ay = ay;
		
		this.vy_boost = 250;
		this.jump_speed = 450;
		this.horiz_speed = 250;
		
		this.time = 0;
		this.ti_x = 0;
		this.ti_y = 0;
		this.t_hit = -1000; //since a player will be red if time_now - time_hit < some number, make sure they're not red to start out by making this v. negative
		
		this.x_collision = 0;
		this.y_collision = 0;
		this.border_y_collision = false;
		
		this.hat = hat;
		this.rightChungus = new Image("rightChungus.png");
		this.leftChungus = new Image("leftChungus.png");
		this.width = width;
		this.chungus_height = (width/rightChungus.getWidth()) * rightChungus.getHeight();
		this.height = .75 * chungus_height; // 0.75 of the image height
		
		this.health = 5;
		this.t_threw_ball = 0;
		this.invincible_timeout = 0.5;
		this.red_color_timeout = 0.5;
		
		this.direction = "right"; //make sure this is consistent with the initialized shootAngle
		
		this.myBall = Optional.empty();
		
		this.shootAngleArray = new double[]{0.05*Math.PI,0.25*Math.PI,0.75*Math.PI,0.95*Math.PI};
		this.shootAngleIndex = 2;
		this.shootAngle = -shootAngleArray[shootAngleIndex];
		this.shootVelocity = 500;
	}
	
	public void bindKeys(KeyCode jumpKey, KeyCode leftKey, KeyCode downKey, KeyCode rightKey, KeyCode rotateLeftKey, KeyCode ballKey, KeyCode rotateRightKey)
	{
		this.jumpKey = jumpKey;
		this.leftKey = leftKey;
		this.downKey = downKey;
		this.rightKey = rightKey;
		this.rotateLeftKey = rotateLeftKey;
		this.ballKey = ballKey;
		this.rotateRightKey = rotateRightKey;
		this.downKeyProcessed = false;
	}
	
	public void giveObjects(ArrayList<Wall> walls, ArrayList<Ball> balls)
	{
		this.walls = walls;
		this.balls = balls;
	}
	
	public void handleKeyPress(KeyCode key)
	{
		//we assume that bindKeys() has been called already
				
		if(key == jumpKey && y_collision == 1) //can only jump if on a platform
		{
			setYVelocity(-jump_speed);
		}
		else if(key == leftKey)
		{
			setXVelocity(-horiz_speed);
		}
		else if(key == rightKey)
		{
			setXVelocity(horiz_speed);
		}
		else if(key == downKey)
		{
			if(!downKeyProcessed) //only process down key once. This gets set to false when we release the down key
			{
				downKeyProcessed = true;
				if(y_collision == 1 && !border_y_collision) //then drop through the platform
				{
					y_collision = 0;
					y += 1;
				}
				else //accelerate downwards
				{
					setYVelocity(vy + ay*(time-ti_y) + vy_boost);
				}
			}
		}
		else if(key == ballKey)
		{
			if(myBall.isEmpty()) //try to grab a ball if we don't have one
			{
				//loop through available balls
				for(Ball b : balls)
				{
					if(b.isDangerous() || b.isHeld()){continue;} //can't pick up dangerous or held balls
					
					//if the center of the ball is within the player, we can pick it up
					if(b.getX() > x && b.getX() < x+width && b.getY() > y && b.getY() < y+height)
					{
						myBall = Optional.of(b);
						b.pickup(this);
						break; //stop searching for balls
					}
				}
			}
			else //we have a ball, shoot it
			{
				shootBall();
			}
		}
		else if(key == rotateLeftKey)
		{
			if(shootAngleIndex < shootAngleArray.length - 1)
			{
				shootAngleIndex++;
				shootAngle = -shootAngleArray[shootAngleIndex];
			}
		}
		else if(key == rotateRightKey)
		{
			if(shootAngleIndex > 0)
			{
				shootAngleIndex--;
				shootAngle = -shootAngleArray[shootAngleIndex];
			}
		}
	}
	
	public void handleKeyRelease(KeyCode key)
	{
		if(key == leftKey || key == rightKey)
		{
			setXVelocity(0);
		}
		else if(key == downKey)
		{
			downKeyProcessed = false;
		}
	}
	
	public void draw(GraphicsContext ctx)
	{
		//make player red if got hit
		if(time - t_hit < red_color_timeout)
		{
			//hooray for copy-pasted code!
			Lighting lighting = new Lighting();
			lighting.setDiffuseConstant(1.0);
			lighting.setSpecularConstant(0.0);
			lighting.setSpecularExponent(0.0);
			lighting.setSurfaceScale(0.0);
			lighting.setLight(new Light.Distant(45, 45, Color.ORANGERED));
			
			ctx.setEffect(lighting);
		}
		
		//draw player
		Image chungus = direction=="left"? leftChungus : rightChungus;
		ctx.drawImage(chungus, x, y-.22*chungus_height, width, chungus_height); //y-.25*chungus_height would place the image perfectly on the platform, but I want to draw it a bit down
		double hat_width = 0.45*width;
		double hat_height = (hat_width/hat.getWidth()) * hat.getHeight();
		double hat_x = x + (direction=="right"? (width-hat_width)*0.6 : (width-hat_width)*0.4);
		ctx.drawImage(hat, hat_x, y-0.2*hat_height, hat_width, hat_height);
		
		//nothing else should be drawn red, reset the red effect
		ctx.setEffect(null);
		
		//draw trajectory direction
		if(myBall.isPresent())
		{
			ctx.setStroke(Color.GRAY);
			ctx.setLineDashes(width/5);
			double ball_x = myBall.get().getX();
			double ball_y = myBall.get().getY();
			ctx.strokeLine(ball_x, ball_y, ball_x + width*Math.cos(shootAngle), ball_y + width*Math.sin(shootAngle));
			
			//draw the ball
			myBall.get().draw(ctx);
		}
	}
	
	//function to update player's position
	public void updatePosition(double time) //time is the current time in seconds, from when the main animation timer started
	{
		this.time = time;
		
		//if velocity is in opposite direction to collision, no more collision
		if(x_collision == 1 && vx < 0 || x_collision == -1 && vx > 0)
		{
			x_collision = 0;
		}
		if(y_collision == 1 && vy < 0 || y_collision == -1 && vy > 0)
		{
			y_collision = 0;
		}
		//if no y_collision, no border_y_collision
		if(y_collision == 0)
		{
			border_y_collision = false;
		}
		
		//define t to be time since most recent velocity reset
		double t_x = time - ti_x;
		double t_y = time - ti_y;
		
		//store previous position
		double prev_x = x;
		double prev_y = y;
		
		//move player if no collisions preventing it
		if(x_collision == 0)
		{
			x = xi + vx*t_x;
		}
		if(y_collision != 1)
		{
			y = yi + vy*t_y + 0.5*ay*t_y*t_y;
		}
		
		//detect wall collision
		boolean foundXCollision = false;
		boolean foundYCollision = false;
		//loop through walls
		for(Wall w : walls)
		{
			//y
			//test collision below
			if( (x+width > w.getX() && x < w.getX()+w.getWidth()) && (prev_y+height <= w.getY() && y+height >= w.getY()) ) //if in right x-range and collide vertically
			{
				foundYCollision = true;
				
				y = w.getY() - height;
				y_collision = 1;
				if(w.isBorderWall()){border_y_collision = true;}
				setYVelocity(0);
			}
			//all other collision tests are dependent on if this is a border wall
			if(w.isBorderWall())
			{
				//test collision above
				if( (x+width > w.getX() && x < w.getX()+w.getWidth()) && (prev_y >= w.getY()+w.getHeight() && y <= w.getY()+w.getHeight()) ) //if in right x-range and collide vertically
				{
					foundYCollision = true;
					
					y = w.getY()+w.getHeight();
					y_collision = -1;
					setYVelocity(0);
				}
				
				//x
				//test collision to right
				if( (y+height > w.getY() && y < w.getY()+w.getHeight()) && (prev_x+width <= w.getX() && x+width >= w.getX()) ) //if in right y-range and collide horizontally
				{
					foundXCollision = true;
					
					x = w.getX() - width;
					x_collision = 1;
					
					//do set velocity stuff without changing the stored velocity
					xi = x;
					ti_x = time;
				}
				//test collision to left
				if( (y+height > w.getY() && y < w.getY()+w.getHeight()) && (prev_x >= w.getX()+w.getWidth() && x <= w.getX()+w.getWidth()) ) //if in right y-range and collide horizontally
				{
					foundXCollision = true;
					
					x = w.getX()+w.getWidth();
					x_collision = -1;
					
					//do set velocity stuff without changing the stored velocity
					xi = x;
					ti_x = time;
				}
			}			
		} //finish looping through walls
		if(!foundXCollision) {x_collision = 0;}
		if(!foundYCollision) {y_collision = 0;}
		
		//check if a ball hit this player
		for(Ball b : balls)
		{
			if(!b.isDangerous()){continue;}
			
			//if the center of the ball is within the player, and time since last hit is bigger than timeout, it hit us
			if(b.getX() > x && b.getX() < x+width && b.getY() > y && b.getY() < y+height && time-t_threw_ball > invincible_timeout)
			{
				health--;
				b.setNotDangerous();
				t_hit = time;
				
				//play chungus chuckle sound
				AudioClip sound = new AudioClip(new File("chuckle.mp3").toURI().toString());
				sound.play();
			}
		}
	}
	
	public void setXVelocity(double vx)//time is the current time in seconds, from when the main animation timer started
	{
		if(x_collision == 1 && vx > 0){return;}
		if(x_collision == -1 && vx < 0){return;}
		
		xi = x;
		this.vx = vx;
		
		ti_x = time;
		
		if(vx < 0){direction = "left";}
		if(vx > 0){direction = "right";}
	}
	
	public void setYVelocity(double vy)//time is the current time in seconds, from when the main animation timer started
	{
		if(y_collision == 1 && vy > 0){return;}
		if(y_collision == -1 && vy < 0){return;}
		
		yi = y;
		this.vy = vy;
		
		ti_y = time;
	}
	
	public void shootBall()
	{
		if(myBall.isEmpty()){return;}
		
		Ball b = myBall.get();
		b.release();
		b.setXVelocity(shootVelocity * Math.cos(shootAngle));
		b.setYVelocity(shootVelocity * Math.sin(shootAngle));
		
		t_threw_ball = time;
		
		myBall = Optional.empty();
	}
	
	public double getX()
	{
		return x;
	}
	public double getY()
	{
		return y;
	}
	public double getWidth()
	{
		return width;
	}
	public double getHeight()
	{
		return height;
	}
	public double getHealth()
	{
		return health;
	}
	public Image getHat()
	{
		return hat;
	}
}

File: Player.js
let server = require("./server");
function getBalls(){return server.getGame().balls;}
function getWalls(){return server.getGame().walls;}

class Player
{

  constructor(name, x, y, width, AY, n_lives, hat_name, hat_aspect_ratio){
    //x, y is initial position
    //width determines size of chungus
    //AY is gravitational acceleration
    //hat_src is name of image file
    //walls and balls are objects stored by the Game object

    this.name = name;

  	this.x = x; //of top-left corner
  	this.xi = x; //initial x position (reset when velocity is reset)
  	this.y = y; //of top-left corner
  	this.yi = y; //initial y position (reset when velocity is reset)
  	this.vx = 0;
  	this.vy = 0;

  	this.AY = AY;

  	this.VY_BOOST = 250; //extra vy to gain from pressing the down key
  	this.JUMP_SPEED = 450;
  	this.HORIZ_SPEED = 250;

  	this.time = 0; //time of most recent updatePosition call
  	this.ti_x = 0; //time of most recent x velocity reset
  	this.ti_y = 0; //time of most recent y velocity reset
  	this.t_hit = -1000; //time the player most recently got hit, used for making them red when hit and for invincibility period
      //since a player will be red if time_now - time_hit < RED_COLOR_TIMEOUT, make sure they're not red to start out by making this very negative
    this.t_died = undefined; //time when lost last life, used for fall-over animation

  	this.x_collision = 0; // -1 means wall to left, 0 means no collision, 1 means wall to right
  	this.y_collision = 0; // -1 means wall above, 0 means no collision, 1 means wall below
  	this.border_y_collision = false; //set to true if the y_collision is with a border wall

  	this.hat_name = hat_name;
    this.hat_aspect_ratio = hat_aspect_ratio; // width / height
  	this.rightChungus_src = "rightChungus.png";
  	this.leftChungus_src = "leftChungus.png";
  	this.width = width; //width of hitbox and of chungus image
  	this.chungus_height = (this.width/1442) * 2618; //image width is 1442, height 2618
  	this.height = 0.75 * this.chungus_height; //height of hitbox

  	this.health = n_lives;
  	this.t_threw_ball = 0; //time when the player last threw a ball; players are invincible for a certain timeout after throwing a ball
  	this.THROWER_INVINCIBLE_TIMEOUT = 0.5; //this is the timeout for invincibility after throwing a ball (so you don't get hit by the ball you threw)
    this.HIT_INVINCIBLE_TIMEOUT = 0.5; //this is the timeout for invincibility after getting hit
  	this.RED_COLOR_TIMEOUT = 0.5; //how long to make the player red after getting hit, in sec

    this.direction = "right"; //"left" or "right"; direction the player is facing
    //make sure this is consistent with the initialized shoot_angle
    this.death_ball_vx = undefined; //vx of the killing ball, at the instant it hit, used for fall-over animation

  	this.ball_index = undefined; //undefined if not holding a ball, otherwise the index in balls where to find the ball
  	this.shoot_angle_array = [0.05*Math.PI,0.25*Math.PI,0.75*Math.PI,0.95*Math.PI]; //stores possible shoot angles, from 0 to pi/2.  These are positive when shooting up as it appears on the canvas.
  	this.shoot_angle_index = 1; //index in shoot angle array to reference
  	this.shoot_angle = -this.shoot_angle_array[this.shoot_angle_index]; //in radians, actual angle of launch, in math coords, not canvas coords
  	this.SHOOT_VELOCITY = 500; //in px/s

    this.JUMP_KEY = "w";
    this.LEFT_KEY = "a";
    this.RIGHT_KEY = "d";
    this.DOWN_KEY = "s";
    this.ROTATE_LEFT_KEY = "i"; //for launch angle changing
    this.ROTATE_RIGHT_KEY = "p";
    this.BALL_KEY = "o";
    this.down_key_processed = false; //keeps track if we processed the first down key press, so that we don't process another one until after keyup
    this.ball_key_processed = false; //same reason as down key
    this.keys_down = []; //list of keys currently down, to continually process (since interactions of multiple keys mean a key held down from a long time ago might not still be sending rapid events)
      //key down an key up handling will maintain this list to be accurate
  }

	handleKeydown(key)
	{
    if(this.health <= 0) return; //dead players can't send inputs, or register a new keydown in the keys_down array

    //maintain this.keys_down
    if(!this.keys_down.includes(key)) this.keys_down.push(key);

    //process keys we only expect behavior from on the instant of keydown/up
    //keys we expect behavior from the duration of being pressed are processed in this.processContinuousKeys()

		if(key == this.DOWN_KEY && !this.down_key_processed) //only process down key once. This gets set to false when we release the down key
		{
				this.down_key_processed = true;
				if(this.y_collision == 1 && !this.border_y_collision) //then drop through the platform
				{
					this.y_collision = 0;
					this.y += 1;
				}
				else //add VY_BOOST to vy to accelerate downwards faster
				{
					this.setYVelocity(this.vy + this.AY*(this.time-this.ti_y) + this.VY_BOOST);
				}
		}
		else if(key == this.BALL_KEY && !this.ball_key_processed) //only process ball key once. This gets set to false when we release the ball key
		{
      this.ball_key_processed = true;
			if(this.ball_index === undefined) //try to grab a ball if we don't have one
			{
				//loop through available balls
				for(let i=0; i<getBalls().length; i++)
				{
          let b = getBalls()[i];
					if(b.isDangerous() || b.isHeld()){continue;} //can't pick up dangerous or held balls

					//if the center of the ball is within the player, we can pick it up
					if(b.x > this.x && b.x < this.x+this.width && b.y > this.y && b.y < this.y+this.height)
					{
						this.ball_index = i;
						b.pickup(this);
						break; //stop searching for balls
					}
				}
			}
			else //we have a ball, shoot it
			{
				this.shootBall();
			}
		}
    else if(key == this.ROTATE_LEFT_KEY)
    {
      if(this.shoot_angle_index < this.shoot_angle_array.length - 1)
      {
        this.shoot_angle_index++;
        this.shoot_angle = -this.shoot_angle_array[this.shoot_angle_index];
      }
    }
    else if(key == this.ROTATE_RIGHT_KEY)
    {
      if(this.shoot_angle_index > 0)
      {
        this.shoot_angle_index--;
        this.shoot_angle = -this.shoot_angle_array[this.shoot_angle_index];
      }
    }
	}

	handleKeyup(key)
	{
    //note: dead players can still register keyups, just not keydowns

    //maintain this.keys_down
    let idx = this.keys_down.indexOf(key);
    if(idx != -1) this.keys_down.splice(idx, 1);


		if(!this.keys_down.includes(this.LEFT_KEY) && !this.keys_down.includes(this.RIGHT_KEY))
		{
			this.setXVelocity(0); //since y velocity is governed by gravity, not resetting like with x velocity
		}
		if(key == this.DOWN_KEY)
		{
			this.down_key_processed = false;
		}
    if(key == this.BALL_KEY){
      this.ball_key_processed = false;
    }
	}


  processContinuousKeys(){
    //continuously process keys currently down regardless of if we're getting events from the client
    //this is called every frame from updatePosition()
    //keys processed here are those we expect behavior from for the duration of being pressed (as opposed to just the instant of keydown/keyup)
    //actions only on the instant of keyup/keydown are processed in handleKeydown/up

    if(this.health <= 0) return;

    this.keys_down.forEach(key => {

      if(key == this.JUMP_KEY && this.y_collision == 1) //can only jump if on a platform
  		{
  			this.setYVelocity(-this.JUMP_SPEED);
  		}
  		else if(key == this.LEFT_KEY)
  		{
  			this.setXVelocity(-this.HORIZ_SPEED);
  		}
  		else if(key == this.RIGHT_KEY)
  		{
  			this.setXVelocity(this.HORIZ_SPEED);
  		}
    });


  }


	//function to update player's position
	updatePosition(time) //time is the current time in seconds, from when the game started
	{
		this.time = time;

		//if velocity is in opposite direction to collision, no more collision
		if(this.x_collision == 1 && this.vx < 0 || this.x_collision == -1 && this.vx > 0)
		{
			this.x_collision = 0;
		}
		if(this.y_collision == 1 && this.vy < 0 || this.y_collision == -1 && this.vy > 0)
		{
			this.y_collision = 0;
		}
		//if no y_collision, no border_y_collision
		if(this.y_collision == 0)
		{
			this.border_y_collision = false;
		}

		//define t to be time since most recent velocity reset
		let t_x = time - this.ti_x;
		let t_y = time - this.ti_y;

		//store previous position
		let prev_x = this.x;
		let prev_y = this.y;

		//move player if no collisions preventing it
		if(this.x_collision == 0)
		{
			this.x = this.xi + this.vx*t_x;
		}
		if(this.y_collision != 1)
		{
			this.y = this.yi + this.vy*t_y + 0.5*this.AY*t_y*t_y;
		}

		//detect wall collision
		let foundXCollision = false;
		let foundYCollision = false;
		//loop through walls
		getWalls().forEach(w => {
			//y
			//test collision below
			if( (this.x+this.width > w.x && this.x < w.x+w.width) && (prev_y+this.height <= w.y && this.y+this.height >= w.y) ) //if in right x-range and collide vertically
			{
				foundYCollision = true;

				this.y = w.y - this.height;
				this.y_collision = 1;
				if(w.border_wall){this.border_y_collision = true;}
				this.setYVelocity(0);
			}
			//all other collision tests are dependent on if this is a border wall
			if(w.border_wall)
			{
				//test collision above
				if( (this.x+this.width > w.x && this.x < w.x+w.width) && (prev_y >= w.y+w.height && this.y <= w.y+w.height) ) //if in right x-range and collide vertically
				{
					foundYCollision = true;

					this.y = w.y+w.height;
					this.y_collision = -1;
					this.setYVelocity(0);
				}

				//x
				//test collision to right
				if( (this.y+this.height > w.y && this.y < w.y+w.height) && (prev_x+this.width <= w.x && this.x+this.width >= w.x) ) //if in right y-range and collide horizontally
				{
					foundXCollision = true;

					this.x = w.x - this.width;
					this.x_collision = 1;

					//do set velocity stuff without changing the stored velocity
					this.xi = this.x;
					this.ti_x = time;
				}
				//test collision to left
				if( (this.y+this.height > w.y && this.y < w.y+w.height) && (prev_x >= w.x+w.width && this.x <= w.x+w.width) ) //if in right y-range and collide horizontally
				{
					foundXCollision = true;

					this.x = w.x+w.width;
					this.x_collision = -1;

					//do set velocity stuff without changing the stored velocity
					this.xi = this.x;
					this.ti_x = time;
				}
			}
		}); //finish looping through walls
		if(!foundXCollision) {this.x_collision = 0;}
		if(!foundYCollision) {this.y_collision = 0;}

		//check if a dangerous ball hit this player
		getBalls().forEach(b => {
			if(!b.isDangerous()){return;}

			//if the center of the ball is within the player, and time since last hit is bigger than timeout, and time since last throw is bigger than timeout, and we're not already dead, it hit us
			if(b.x > this.x && b.x < this.x+this.width && b.y > this.y && b.y < this.y+this.height
        && time-this.t_hit > this.HIT_INVINCIBLE_TIMEOUT && time-this.t_threw_ball > this.THROWER_INVINCIBLE_TIMEOUT
        && this.health > 0)
			{
				this.health = Math.max(this.health - 1, 0);
				b.setNotDangerous();
				this.t_hit = time;

        //check if died
        if(this.health <= 0){
          this.t_died = time;
          this.death_ball_vx = b.vx;
          this.setXVelocity(0); //stop key input, y velocity should continue but can only be affected by gravity now
          this.dropBall();
        }
			}
		});


    this.processContinuousKeys();
	}

	setXVelocity(vx)//time is the current time in seconds, from when the main animation timer started
	{
		if(this.x_collision == 1 && vx > 0){return;}
		if(this.x_collision == -1 && vx < 0){return;}

		this.xi = this.x;
		this.vx = vx;

		this.ti_x = this.time;

		if(vx < 0){this.direction = "left";}
		if(vx > 0){this.direction = "right";}
	}

	setYVelocity(vy)//time is the current time in seconds, from when the main animation timer started
	{
		if(this.y_collision == 1 && vy > 0){return;}
		if(this.y_collision == -1 && vy < 0){return;}

		this.yi = this.y;
		this.vy = vy;

		this.ti_y = this.time;
	}

	shootBall()
	{
		if(this.ball_index === undefined){return;}

		let b = getBalls()[this.ball_index];
		b.release();
		b.setXVelocity(this.SHOOT_VELOCITY * Math.cos(this.shoot_angle));
		b.setYVelocity(this.SHOOT_VELOCITY * Math.sin(this.shoot_angle));
		//TODO consider adding player's velocity to shoot velocity

		this.t_threw_ball = this.time;

		this.ball_index = undefined;
	}

  dropBall(){
    //called when a player dies if holding a ball so they don't hoard a ball
    if(this.ball_index === undefined){return;}

    let b = getBalls()[this.ball_index];
    b.release();
    b.setNotDangerous();
    b.setXVelocity(0);
    b.setYVelocity(0);

    this.ball_index = undefined;
  }
}


exports.Player = Player;
